src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,1052 @@
+    }
+  }
+}
+
+/**
+ * Applies a separable linear filter to an Image
+ * @param InOutImage Input/Output Contains the input Image, The final filtered Image is
+ *          stored in this param.
+ * @param kernel Input Kernel vector to be applied on Image.
+ * @param radius amount, the Image should be padded before applying filter.
+ */
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+SepFilter2D(CubeType &InOutImage, const arma::vec& kernel, const size_t radius)
+{
+  CubeType OutImage;
+  this->CopyMakeBorder(InOutImage, radius, radius, radius, radius, OutImage);
+
+  arma::vec row_res, col_res;
+  // reverse InOutImage and OutImage to avoid making an extra matrix.
+  // InImage is renamed to InOutImage in this function for this reason only.  
+  MatType k_mat = kernel * kernel.t();
+  for(size_t k = 0; k < OutImage.n_slices; ++k)
+  {
+    for(size_t j = radius; j < OutImage.n_cols - radius; ++j)
+    {
+      for(size_t i = radius; i < OutImage.n_rows - radius; ++i)
+      {
+        InOutImage(i - radius, j - radius, k) = 
+            arma::accu(OutImage.slice(k)\
+            .submat(i - radius, j - radius,\
+              i + radius, j + radius) % k_mat);
+      }
+    }
+  }
+
+}
+
+/**
+ * Applies a triangle filter on an Image.
+ * @param InImage Input/Output Image on which filter is applied.
+ * @param radius Decides the size of kernel to be applied on Image.
+ */
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+ConvTriangle(CubeType &InImage, const size_t radius)
+{
+  if (radius == 0)
+  {
+    //nothing to do
+  }
+  else if (radius <= 1)
+  {
+    const double p = 12.0 / radius / (radius + 2) - 2;
+    arma::vec kernel = {1, p, 1};
+    kernel /= (p + 2);
+    
+    this->SepFilter2D(InImage, kernel, radius);
+  }
+  else
+  {
+    const size_t len = 2 * radius + 1;
+    arma::vec kernel(len);
+    for( size_t i = 0; i < radius; ++i)
+      kernel(i) = i + 1;
+    
+    kernel(radius) = radius + 1;
+    
+    size_t r = radius;
+    for( size_t i = radius + 1; i < len; ++i)
+      kernel(i) = r--;
+
+    kernel /= std::pow(radius + 1, 2);
+    this->SepFilter2D(InImage, kernel, radius);
+  }
+}
+
+//just a helper function, can't use it for anything else
+//finds max numbers on cube axis and returns max values,
+// also stores the locations of max values in Location
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+MaxAndLoc(CubeType& mag, arma::umat& Location, MatType& MaxVal) const
+{
+  /*Vectorize this function after prototype works*/
+  MaxVal = MatType(Location.n_rows, Location.n_cols);
+  for(size_t i = 0; i < mag.n_rows; ++i)
+  {
+    for(size_t j = 0; j < mag.n_cols; ++j)
+    {
+      /*can use -infinity here*/
+      double max =  std::numeric_limits<double>::min();
+      for(size_t k = 0; k < mag.n_slices; ++k)
+      {
+        if(mag(i, j, k) > max)
```

Comment:
I figured out why it doesn't work on my machine. The check: 

``if(mag(i, j, k) > max)``

doesn't work if mag is [0 0 0] because 0 < double max =  std::numeric_limits<double>::min();.  Since the the location is undefined, we could probably run into an index out of bounds error.

As pointed out here: https://github.com/mlpack/mlpack/pull/696#discussion_r67073948 we should use -DBL_MAX or just use the first value of mag as starting point.